File: go/src/onde/connection.go
package onde

import (
	"encoding/json"
	"fmt"
	"gopkg.in/igm/sockjs-go.v2/sockjs"
	"log"
	. "onde/api"
	"onde/card"
	"onde/search"
	"strings"
	"onde/solr"
)

type Connection struct {
	user       solr.JsonObject
	sock       sockjs.Session
	cardSubs    map[int]*card.Card // subId -> Card
	searchSubs map[string]*search.Search  // query -> Search
}

func sockHandler(sock sockjs.Session) {
	log.Printf("new connection: %s", sock.ID())

	var conn *Connection
	var err error
	for {
		var msg string
		if msg, err = sock.Recv(); err == nil {
			var req Req
			err := json.NewDecoder(strings.NewReader(msg)).Decode(&req)
			if err != nil {
				log.Printf("failed to parse req: %s", err)
				break
			}

			switch req.Type {
			case MsgLogin:
				userId := req.Login.UserId
				user, err := FindUser(userId)
				if err != nil {
					ErrorRsp{Msg: fmt.Sprintf("Invalid user id: %s", userId)}.Send(sock)
					continue
				}
				pass := user.GetString("prop_pass")
				log.Printf("pass: %v", pass)
				if pass != nil && req.Login.Password != *pass {
					ErrorRsp{Msg: fmt.Sprintf("Incorrect password for user: %s", userId)}.Send(sock)
					continue
				}
				conn = newConnection(user, sock)
				LoginRsp{UserId: req.Login.UserId, ConnId: conn.Id()}.Send(sock)

			case MsgSubscribeCard:
				if conn.validate(sock) {
					conn.handleSubscribeCard(req.SubscribeCard)
				}

			case MsgUnsubscribeCard:
				if conn.validate(sock) {
					conn.handleUnsubscribeCard(req.UnsubscribeCard)
				}

			case MsgRevise:
				if conn.validate(sock) {
					conn.handleRevise(req.Revise)
				}

			case MsgSubscribeSearch:
				if conn.validate(sock) {
					conn.handleSubscribeSearch(req.SubscribeSearch)
				}

			case MsgUnsubscribeSearch:
				if conn.validate(sock) {
					conn.handleUnsubscribeSearch(req.UnsubscribeSearch)
				}

			case MsgCreateCard:
				if conn.validate(sock) {
					conn.handleCreateCard(req.CreateCard)
				}
			}

			continue
		}
		break
	}

	if conn != nil {
		conn.cleanupSubs()
	}

	log.Printf("lost connection %s: %s", sock.ID(), err)
}

func (conn *Connection) Id() string {
	return conn.sock.ID()
}

func (conn *Connection) validate(sock sockjs.Session) bool {
	if conn == nil {
		ErrorRsp{Msg: "no connection"}.Send(sock)
		return false
	}
	return true
}

func (conn *Connection) handleSubscribeCard(req *SubscribeCardReq) {
	if _, exists := conn.cardSubs[req.SubId]; exists {
		ErrorRsp{Msg: fmt.Sprintf("double subscribe subid %d", req.SubId)}.Send(conn.sock)
		return
	}

	card, err := card.Subscribe(req.CardId, conn.Id(), req.SubId, conn.sock)
	if err != nil {
		ErrorRsp{Msg: fmt.Sprintf("no such card: %s", req.CardId)}.Send(conn.sock)
		return
	}
	conn.cardSubs[req.SubId] = card

	SubscribeCardRsp{
		CardId: req.CardId,
		SubId: req.SubId,
		Rev:   card.Rev(),
		Props: card.Props(),
	}.Send(conn.sock)
}

func (conn *Connection) handleUnsubscribeCard(req *UnsubscribeCardReq) {
	card, exists := conn.cardSubs[req.SubId]
	if !exists {
		ErrorRsp{Msg: fmt.Sprintf("error unsubscribing subid %d: no subscription found", req.SubId)}.Send(conn.sock)
		return
	}

	delete(conn.cardSubs, req.SubId)
	card.Unsubscribe(conn.Id(), req.SubId)
	UnsubscribeCardRsp{SubId: req.SubId}.Send(conn.sock)
}

func (conn *Connection) handleRevise(req *ReviseReq) {
	card, exists := conn.cardSubs[req.SubId]
	if !exists {
		ErrorRsp{Msg: fmt.Sprintf("error revising card %s - not subscribed", req.CardId)}.Send(conn.sock)
		return
	}
	card.Revise(req.ConnId, req.SubId, req.Rev, req.Change)
}

func (conn *Connection) handleSubscribeSearch(req *SubscribeSearchReq) {
	if _, exists := conn.searchSubs[req.Query]; exists {
		ErrorRsp{Msg: fmt.Sprintf("double subscribe: %s", req.Query)}.Send(conn.sock)
		return
	}

	search, err := search.Subscribe(req.Query, conn.Id(), conn.sock)
	if err != nil {
		ErrorRsp{Msg: fmt.Sprintf("unable to subscribe to search: %s", req.Query)}.Send(conn.sock)
		return
	}

	conn.searchSubs[req.Query] = search
	SubscribeSearchRsp{
		Query: req.Query,
	}.Send(conn.sock)
}

func (conn *Connection) handleUnsubscribeSearch(req *UnsubscribeSearchReq) {
	s, exists := conn.searchSubs[req.Query]
	if !exists {
		ErrorRsp{Msg: fmt.Sprintf("error unsubscribing search %s: no subscription found", req.Query)}.Send(conn.sock)
		return
	}

	delete(conn.searchSubs, req.Query)
	s.Unsubscribe(conn.Id())
	UnsubscribeSearchRsp{Query: req.Query}.Send(conn.sock)
}

func (conn *Connection) handleCreateCard(req *CreateCardReq) {
	cardId, err := card.Create(conn.Id(), req.Props)
	if err != nil {
		ErrorRsp{Msg: fmt.Sprintf("error creating card: %s", err)}.Send(conn.sock)
		return
	}
	CreateCardRsp{CreateId: req.CreateId, CardId: cardId}.Send(conn.sock)
}

func (conn *Connection) cleanupSubs() {
	// Remove this connection's subscriptions from their cards.
	// Don't bother clearing conn.*Subs, because it won't be reused
	for subId, card := range conn.cardSubs {
		card.Unsubscribe(conn.Id(), subId)
	}
	for _, s := range conn.searchSubs {
		s.Unsubscribe(conn.Id())
	}
}

func newConnection(user solr.JsonObject, sock sockjs.Session) *Connection {
	return &Connection{
		user:       user,
		sock:       sock,
		cardSubs:    make(map[int]*card.Card),
		searchSubs: make(map[string]*search.Search),
	}
}

File: ts/connection.ts
/// <reference path="lib/sockjs.d.ts" />

// TODO:
// - Implement outgoing message queue.
// - Re-establish connection automatically.
// - Track acknowledged requests.
module onde {

  var LOG_MESSAGES = true;

  export class CardSubscription {
    _subId: number;

    constructor(
        private _conn: Connection,
        public cardId: string,
        public _onsubscribe: (rsp: SubscribeCardRsp) => void,
        public _onrevision: (rsp: ReviseRsp) => void,
        public _onack: (rsp: ReviseRsp) => void) {
      this._subId = ++_conn._curSubId;
    }

    revise(rev: number, change: Change) {
      var req: Req = {
        Type: MsgRevise,
        Revise: { ConnId: this._conn.connId(), SubId: this._subId, CardId: this.cardId, Rev: rev, Change: change }
      };
      this._conn._send(req);
    }

    unsubscribe() {
      var req: Req = {
        Type: MsgUnsubscribeCard,
        UnsubscribeCard: { SubId: this._subId }
      };
      this._conn._send(req);
    }
  }

  export class SearchSubscription {
    constructor(
        private _conn: Connection,
        public query: string,
        public _onsearchresults: (rsp: SearchResultsRsp) => void) { }

    unsubscribe() {
      var subs = this._conn._searchSubs[this.query];
      subs.splice(subs.indexOf(this), 1);
      if (subs.length == 0) {
        var req: Req = {
          Type: MsgUnsubscribeSearch,
          UnsubscribeSearch: { Query: this.query }
        };
        this._conn._send(req);
        delete this._conn._searchSubs[this.query];
      }
    }
  }

  export class Connection {
    private _sock: SockJS;
    private _connId: string;
    private _onCreates: {[createId: number]: (rsp: CreateCardRsp) => void} = {};

    _cardSubs: {[key: string]: CardSubscription} = {};
    _searchSubs: {[query: string]: SearchSubscription[]} = {};
    _curSubId = 0;
    _curCreateId = 0;

    onOpen: () => void;
    onClose: () => void;
    onLogin: () => void;

    constructor(private _ctx: Context) {
    }

    connect() {
      this._sock = new SockJS(this.getOrigin() + "/sock", null, {
        debug: true
      });

      this._sock.onopen = () => {
        if (this.onOpen) {
          this.onOpen();
        }
      };
      this._sock.onclose = () => {
        this._sock = null;
        this.connId = null;
        if (this.onClose) {
          this.onClose();
        }
      };
      this._sock.onmessage = (e: SJSMessageEvent) => { this.onMessage(e); };
    }

    connId(): string {
      return this._connId;
    }

    login(userId: string) {
      var req: Req = {
        Type: MsgLogin,
        Login: { UserId: userId }
      };
      this._send(req);
    }

    subscribeCard(cardId: string, onSubscribe: (rsp: SubscribeCardRsp) => void, onRevision: (rsp: ReviseRsp) => void, onAck: (rsp: ReviseRsp) => void): CardSubscription {
      var sub = new CardSubscription(this, cardId, onSubscribe, onRevision, onAck);
      this._cardSubs[cardSubKey(cardId, sub._subId)] = sub;

      var req: Req = {
        Type: MsgSubscribeCard,
        SubscribeCard: { CardId: cardId, SubId: sub._subId }
      };
      this._send(req);
      return sub;
    }

    subscribeSearch(query: string, onSearchResults: (rsp: SearchResultsRsp) => void): SearchSubscription {
      if (!(query in this._searchSubs)) {
        this._searchSubs[query] = [];
        var req: Req = {
          Type: MsgSubscribeSearch,
          SubscribeSearch: { Query: query }
        };
        this._send(req);
      }

      var sub = new SearchSubscription(this, query, onSearchResults);
      this._searchSubs[query].push(sub);
      return sub;
    }

    createCard(props: {[prop: string]: string}, onCreated: (rsp: CreateCardRsp) => void) {
      var id = ++this._curCreateId;
      this._onCreates[id] = onCreated;
      var req: Req = {
        Type: MsgCreateCard,
        CreateCard: {
          CreateId: id,
          Props: props
        }
      };
      this._send(req);
    }

    _send(req: Req) {
      if (LOG_MESSAGES) {
        this._ctx.log(req);
      }
      this._sock.send(JSON.stringify(req));
    }

    private getOrigin(): string {
      return location.protocol + "//" + location.hostname + (location.port ? (":" + location.port) : "");
    }

    private handleLogin(rsp: LoginRsp) {
      this._connId = rsp.ConnId;
      if (this.onLogin) {
        this.onLogin();
      }
    }

    private handleSubscribeCard(rsp: SubscribeCardRsp) {
      var sub = this._cardSubs[cardSubKey(rsp.CardId, rsp.SubId)];
      sub._onsubscribe(rsp);
    }

    private handleRevise(rsp: ReviseRsp) {
      for (var i = 0; i < rsp.SubIds.length; ++i) {
        var sub = this._cardSubs[cardSubKey(rsp.CardId, rsp.SubIds[0])];
        if (!sub) {
          this._ctx.log("got results for card " + rsp.CardId + " with no local subscription");
          continue;
        }

        if ((rsp.OrigConnId == this._connId) && (rsp.OrigSubId == sub._subId)) {
          sub._onack(rsp);
        } else {
          sub._onrevision(rsp);
        }
      }
    }

    private handleSearchResults(rsp: SearchResultsRsp) {
      var subs = this._searchSubs[rsp.Query];
      if (!subs) {
        this._ctx.log("got results for search " + rsp.Query + " with no local subscription");
        return;
      }

      for (var i = 0; i < subs.length; ++i) {
        subs[i]._onsearchresults(rsp);
      }
    }

    private handleCreateCard(rsp: CreateCardRsp) {
      var onCreate = this._onCreates[rsp.CreateId];
      if (!onCreate) {
        this._ctx.log("got unmatched create response " + rsp.CreateId);
        return;
      }

      delete this._onCreates[rsp.CreateId];
      onCreate(rsp);
    }

    private onMessage(e: SJSMessageEvent) {
      var rsp = <Rsp>JSON.parse(e.data);
      if (LOG_MESSAGES) {
        this._ctx.log(rsp);
      }
      switch (rsp.Type) {
        case MsgLogin:
          this.handleLogin(rsp.Login);
          break;

        case MsgSubscribeCard:
          this.handleSubscribeCard(rsp.SubscribeCard);
          break;

        case MsgUnsubscribeCard:
          // TODO
          break;

        case MsgRevise:
          this.handleRevise(rsp.Revise);
          break;

        case MsgSubscribeSearch:
          // TODO
          break;

        case MsgSearchResults:
          this.handleSearchResults(rsp.SearchResults);
          break;

        case MsgUnsubscribeSearch:
          // TODO
          break;

        case MsgCreateCard:
          this.handleCreateCard(rsp.CreateCard);
          break;

        case MsgError:
          this._ctx.log(rsp.Error.Msg);
          break;
      }
    }
  }

  function cardSubKey(cardId: string, subId: number): string {
    return cardId + ":" + subId;
  }
}
